# Funnel Analytics for Metro car


In [2]:
import numpy as np
import random 
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd
from sklearn.linear_model import LinearRegression
import plotly.express as px
import plotly.graph_objects as go

Metrocar is a ride-hailing service (similar to Uber/Lyft). Metrocar's business
concept is built upon a platform that links riders and drivers via a
smartphone app. Metrocar is anmediator between passengers and drivers,
offering a user-friendly platform for them to connect and Make the
ride-hailing procedure easier.


In [3]:
 funnel1= pd.read_csv('C://Users//mande//OneDrive//Desktop//metro_sql_submission//submissions/query_results-funnel.csv')

In [4]:
df= pd.DataFrame( funnel1)
df

funnel_step     funnel_name  user_count      lag    diff  conversion_rate  \
0            1    app_download       23608      NaN     NaN              NaN   
1            2         sign_up       17623  23608.0  5985.0           0.7465   
2            3  ride_requested       12406  17623.0  5217.0           0.7040   
3            4   ride_accepted       12278  12406.0   128.0           0.9897   
4            5  ride_completed        6233  12278.0  6045.0           0.5077   
5            6         payment        6233   6233.0     0.0           1.0000   
6            7          review        4348   6233.0  1885.0           0.6976   

   dropoff_percent  
0              NaN  
1           0.2535  
2           0.2960  
3           0.0103  
4           0.4923  
5           0.0000  
6           0.3024

In [13]:
# Assuming your CSV file contains columns: 'funnel_step' and 'user_count'
steps = df['funnel_step']
values = df['user_count']

# Define a color sequence
colors = ['peru', 'darkorange', 'purple', 'darkslateblue']

# Create the trace for the funnel chart with defined colors
fig = go.Figure(go.Funnel(
    y=steps,
    x=values,
    textinfo="value+percent initial",
    marker={"color": colors}
))

# Update layout
fig.update_layout(title="Funnel Chart",
                  xaxis_title="Number of Users",
                  yaxis_title="Steps")

# Show the plot
fig.show()

Funnel Analysis Metrics Breakdown
Funnel Step: Represents the specific stage in the customer journey.
Funnel Name: Identifies the corresponding step in the customer journey.
User Count: Total count of users at each stage within the funnel.
Lag: Indicates the user count difference between the current and previous
steps, reflecting user progression.
Difference (Diff): Represents the absolute variance in user count
compared to the initial step (app_download), signifying the gain or loss of
users at each stage.
Conversion Rate: Calculated as the ratio of users moving from one step to
the next. For instance, for "sign_up," the conversion rate at 0.7465 denotes
that 74.65% of app downloaders proceeded to sign up.
Drop-off Percent: Complementary to the conversion rate, indicating the
rate at which users do not progress to the subsequent step. For example,
for "sign_up," the drop-off percent stands at 0.2535, signifying that 25.35%
of users who downloaded the app did not complete the sign-up process.
Key Insights from Customer Funnel Analysis:
Initial Engagement: Commenced with 23,608 users downloading the app,
marking the start of the funnel.
Conversion at Sign-Up: Demonstrated a robust conversion rate of
74.65% as app downloaders progressed to sign up.
Ride Request Transition: Showcased a 70.40% conversion rate from
sign-up to ride request, signifying consistent user engagement.
Ride Acceptance: Impressed with a high 98.97% conversion rate from ride
request to ride acceptance, indicating minimal drop-off.
Ride Completion Challenge: Encountered a notable drop with only
50.77% of users who accepted rides proceeding to complete them,
suggesting a critical point for improvement.
Seamless Payment Process: Achieved a perfect conversion rate of 100%
from ride completion to payment, indicating a frictionless transactional
experience.
Review Engagement: Showed a 69.76% conversion rate from payment to
review, though with a significant drop-off rate of 30.24%, highlighting an
area for feedback enhancement.

In [5]:
 funnel2= pd.read_csv('C://Users//mande//OneDrive//Desktop//metro_sql_submission/metrocar_funnel_by_USER_metrocar_funnel_by_USER.csv')

In [14]:
# Create a pie chart using Plotly Express
fig_pie = px.pie(funnel2, values='Users Count', names='Platform', title='Downloads by Platform')
fig_pie.show()

Key Insights:
● Metrocar currently caters to three platforms: iOS, Android, and Web. iOS holds
the majority share at 60.53%, followed by Android with 29.38%.
● Analysis via Tableau funnel visualization based on platform-specific filters
reveals no significant variance in conversion and dropoff rates across platforms

In [7]:
 funnel_age= pd.read_csv('C://Users//mande//OneDrive//Desktop//metro_sql_submission//submissions/query_results-age.csv')
funnel_age

age_range  signups  total_signups  pct_of_signups
0     18-24     1865          17623          0.1058
1     25-34     3447          17623          0.1956
2     35-44     5181          17623          0.2940
3     45-54     1826          17623          0.1036
4   Unknown     5304          17623          0.3010

In [9]:
funnel_age= pd.read_csv('C://Users//mande//OneDrive//Desktop//metro_sql_submission//submissions/query_results-funnel.csv')
funnel_age

funnel_step     funnel_name  user_count      lag    diff  conversion_rate  \
0            1    app_download       23608      NaN     NaN              NaN   
1            2         sign_up       17623  23608.0  5985.0           0.7465   
2            3  ride_requested       12406  17623.0  5217.0           0.7040   
3            4   ride_accepted       12278  12406.0   128.0           0.9897   
4            5  ride_completed        6233  12278.0  6045.0           0.5077   
5            6         payment        6233   6233.0     0.0           1.0000   
6            7          review        4348   6233.0  1885.0           0.6976   

   dropoff_percent  
0              NaN  
1           0.2535  
2           0.2960  
3           0.0103  
4           0.4923  
5           0.0000  
6           0.3024

In [10]:

# Create a bar chart for conversion rate including the funnel names
fig = px.bar(funnel_age, x='funnel_step', y='conversion_rate', color='funnel_name',
             labels={'funnel_step': 'Funnel Step', 'conversion_rate': 'Conversion Rate'},
             title='Conversion Rate for Different Funnel Steps',
             hover_data=['user_count', 'lag', 'diff', 'dropoff_percent'])

# Update layout
fig.update_layout(xaxis_title='Funnel Steps', yaxis_title='Conversion Rate')

# Show the plot
fig.show()

Age Group Performance in the Funnel:
Insights:
● Age Distribution: The "35-44" age range stands out, constituting 29.40% of
signups, followed by "25-34" at 19.56%. The "Unknown" category, encompassing
30.10% of signups, likely comprises users who didn't disclose their age.
● Lower Signup Percentages: "18-24" and "45-54" age groups exhibit lower signup
rates, accounting for 10.58% and 10.36%, respectively.
● Unrevealed Age Data: The "Unknown" category, representing 30.10% of signups,
potentially holds crucial insights if users disclose their age information.
● Funnel Analysis by Age Range: Tableau funnel visualization suggests no notable
variance in conversion and dropoff rates across age ranges post signup.

In [11]:
funnel_full= pd.read_csv('C://Users//mande//OneDrive//Desktop//metro_sql_submission//submissions/query_results-2023-10-10_44039.csv')
funnel_full

app_download_key platform          download_ts  \
0      089c3299cb7cf4cf752711d29d916ac1      ios  2021-12-11 05:05:01   
1      17f8d5dd82056d27105f9c28be15668a  android  2021-05-24 16:35:48   
2      ea9194aa4c722172e59183a968613a7b      ios  2021-09-12 15:55:14   
3      cecf8d52ce305f077d5c2fffd22f9b10  android  2021-09-11 14:20:59   
4      8221632857a6aaa5032752b5c7be31f0      ios  2021-06-16 10:01:11   
...                                 ...      ...                  ...   
49995  9cf532cd4d8f019781073edf3b870fe4      ios  2021-02-19 05:27:59   
49996  1924fff4dfc20dd6f0aa8a028ed2c16c      ios  2021-03-21 19:16:50   
49997  b67f5e4bc6937f88c95d28214b9b9003  android  2021-02-25 00:11:22   
49998  31ed231bfa0311edbf80f1a5ad8a1f16      web  2021-09-21 05:51:40   
49999  ab552c602c2728a659b88dcfebe695c7      ios  2021-04-06 21:34:28   

                             session_id  user_id age_range  \
0      089c3299cb7cf4cf752711d29d916ac1   116687     25-34   
1      17f8d5dd82056d27105f9c28be15668a   106866     45-54   
2      ea9194aa4c722172e59183a968613a7b   112324   Unknown   
3      cecf8d52ce305f077d5c2fffd22f9b10   112281     35-44   
4      8221632857a6aaa5032752b5c7be31f0   108020     35-44   
...                                 ...      ...       ...   
49995  9cf532cd4d8f019781073edf3b870fe4   102410   Unknown   
49996  1924fff4dfc20dd6f0aa8a028ed2c16c   103826   Unknown   
49997  b67f5e4bc6937f88c95d28214b9b9003   102688     35-44   
49998  31ed231bfa0311edbf80f1a5ad8a1f16   112730   Unknown   
49999  ab552c602c2728a659b88dcfebe695c7   104599     35-44   

                 signup_ts  ride_id  driver_id           request_ts  ...  \
0      2021-12-11 11:18:01  3000288        NaN  2021-12-12 08:57:00  ...   
1      2021-05-25 19:32:48  3000028        NaN  2021-05-26 08:15:00  ...   
2      2021-09-12 23:11:14  3000029   106121.0  2021-09-13 19:56:00  ...   
3      2021-09-12 13:08:59  3000035        NaN  2021-09-13 09:28:00  ...   
4      2021-06-16 23:28:11  3000040   102835.0  2021-06-17 09:29:00  ...   
...                    ...      ...        ...                  ...  ...   
49995  2021-02-20 22:46:59  3281490   109649.0  2022-01-19 19:38:00  ...   
49996  2021-03-23 05:56:50  3281500   107868.0  2022-01-14 17:58:00  ...   
49997  2021-02-25 01:46:22  3281504        NaN  2021-04-01 17:44:00  ...   
49998  2021-09-22 17:33:40  3281510        NaN  2021-10-17 17:42:00  ...   
49999  2021-04-07 13:20:28  3281512        NaN  2021-05-18 11:07:00  ...   

                 pickup_ts           dropoff_ts            cancel_ts  \
0                      NaN                  NaN  2021-12-12 09:05:00   
1                      NaN                  NaN  2021-05-26 08:19:00   
2      2021-09-13 20:19:00  2021-09-13 21:26:00                  NaN   
3                      NaN                  NaN  2021-09-13 09:47:00   
4      2021-06-17 09:48:00  2021-06-17 11:14:00                  NaN   
...                    ...                  ...                  ...   
49995  2022-01-19 19:50:00  2022-01-19 20:58:00                  NaN   
49996                  NaN                  NaN  2022-01-14 18:23:00   
49997                  NaN                  NaN  2021-04-01 17:48:00   
49998                  NaN                  NaN  2021-10-17 17:52:00   
49999                  NaN                  NaN  2021-05-18 11:21:00   

      transaction_id purchase_amount_usd charge_status       transaction_ts  \
0                NaN                 NaN           NaN                  NaN   
1                NaN                 NaN           NaN                  NaN   
2         10000018.0               13.04       Decline  2021-09-13 21:26:00   
3                NaN                 NaN           NaN                  NaN   
4         10000024.0               10.20      Approved  2021-06-17 11:14:00   
...              ...                 ...           ...                  ...   
49995     10163386.0               16.60      Approved  2022-01-19 20:

In [12]:

# Convert relevant timestamp columns to datetime objects
date_columns = ['request_ts', 'pickup_ts', 'dropoff_ts', 'cancel_ts']
for column in date_columns:
    funnel_full[column] = pd.to_datetime(funnel_full[column])

# Extract the hour for canceled and completed rides
funnel_full['cancel_hour'] = funnel_full['cancel_ts'].dt.hour
funnel_full['completed_hour'] = funnel_full['dropoff_ts'].dt.hour

# Count canceled rides per hour
cancelled_rides_per_hour = funnel_full.groupby('cancel_hour').size()

# Count completed rides per hour
completed_rides_per_hour = funnel_full.groupby('completed_hour').size()

# Create interactive line plots for canceled and completed rides per hour using Plotly
cancelled_trace = go.Scatter(x=cancelled_rides_per_hour.index, y=cancelled_rides_per_hour.values, mode='lines', name='Canceled Rides', line=dict(color='red'))
completed_trace = go.Scatter(x=completed_rides_per_hour.index, y=completed_rides_per_hour.values, mode='lines', name='Completed Rides', line=dict(color='green'))

# Define layout
layout = go.Layout(title='Canceled and Completed Rides per Hour', xaxis=dict(title='Hour of the Day'), yaxis=dict(title='Count of Rides'), legend=dict(x=0, y=1.0))

# Create figure and plot
fig = go.Figure(data=[cancelled_trace, completed_trace], layout=layout)
fig.show()

Recommendations to Drive Revenue Growth and Enhance Customer Experience:
● Optimize Sign-Up Process: Enhance the transition from "app_download" to
"sign_up" by streamlining the process to alleviate the 25.35% drop-off. Focus on
minimizing sign-up complexities to enhance user retention.
● Enhance Ride Acceptance: Investigate and refine the user journey from
"ride_requested" to "ride_accepted" to reduce the 10.03% drop-off. Prioritize
improvements that facilitate seamless ride acceptance.
● Address "ride_accepted" to "ride_completed" Drop-Off: Analyze the substantial
49.23% drop-off between these stages. Conduct thorough supply-side analysis,
encourage comprehensive user reviews, leverage sentiment analysis for deeper
insights, and align demand-side and supply-side funnel analysis for holistic
improvements.
● Platform Equity: Sustain consistent attention and support across iOS, Android,
and Web due to similar conversion and drop-off rates to ensure uniform user
experiences.
● Strategic Marketing Allocation: Distribute marketing resources proportionate to
user platform shares, focusing particularly on the considerable difference in user
volumes between iOS and Android.
● Targeted Marketing and Enhanced Experience: Channel marketing efforts and
experience enhancements toward the "35-44" and "25-34" age brackets,
optimizing user journeys based on their higher sign-up percentages.
● User Segmentation and Personalization: Leverage user data, including canceled
rides, and encourage age information during sign-up to tailor experiences.
Highlight incentives or benefits to prompt data sharing and provide personalized
journeys.
● Strategic Surge Pricing: Implement surge pricing during peak hours (8 AM - 10 AM
and 4 PM - 8 PM) to effectively manage revenue streams during heightened
demand periods.
